File: src/src/lexer.cpp
#include "ollang.h"
#include <unordered_set>
#include <cctype>

char Lexer::current() const {
    return pos < source.length() ? source[pos] : '\0';
}

char Lexer::peek(size_t offset) const {
    return pos + offset < source.length() ? source[pos + offset] : '\0';
}

void Lexer::advance() {
    if (current() == '\n') {
        line++;
        col = 1;
    }
    else {
        col++;
    }
    pos++;
}

void Lexer::skipWhitespace() {
    while (isspace(current())) advance();
}

void Lexer::skipComment() {
    if (current() == '#') {
        while (current() != '\n' && current() != '\0') advance();
    }
}

double Lexer::readNumber() {
    std::string num;
    while (isdigit(current()) || current() == '.') {
        num += current();
        advance();
    }
    return std::stod(num);
}

std::string Lexer::readString() {
    std::string str;
    advance(); // Skip opening quote
    while (current() != '"' && current() != '\0') {
        if (current() == '\\') {
            advance();
            switch (current()) {
            case 'n': str += '\n'; break;
            case 't': str += '\t'; break;
            case 'r': str += '\r'; break;
            case '"': str += '"'; break;
            case '\\': str += '\\'; break;
            default: str += current(); break;
            }
        }
        else {
            str += current();
        }
        advance();
    }
    advance(); // Skip closing quote
    return str;
}

std::string Lexer::readIdentifier() {
    std::string id;
    while (isalnum(current()) || current() == '_') {
        id += current();
        advance();
    }
    return id;
}

Lexer::Lexer(const std::string& src) : source(src) {}

std::vector<Token> Lexer::tokenize() {
    std::vector<Token> tokens;
    std::unordered_set<std::string> keywords = {
        "func", "if", "else", "while", "for", "in", "return",
        "true", "false", "null", "alloc", "free", "read", "write", "syscall",
        "import", "ImportDLL", "try", "catch", "async", "await", "throw", "namespace",
        "process", "module", "inject", "hook", "scan", "window", "thread",
        "httpGet", "httpPost", "httpPut", "httpDelete",
        "base64Encode", "base64Decode", "base64EncodeFile", "base64DecodeFile",
        "base64UrlEncode", "base64UrlDecode", "UUID", // Uniquie User Identifier
        "checkapp", "getapppid", "waitforapp", "waitforappclose", "killapp", "startapp",
        "getAvailableMemory"
    };

    while (current() != '\0') {
        skipWhitespace();
        skipComment();
        if (current() == '\0') break;

        int currentLine = line;
        int currentCol = col;
        char c = current();

        // Handle numbers
        if (isdigit(c)) {
            double num = readNumber();
            tokens.push_back(Token("number", num, currentLine, currentCol));
        }
        // Handle strings
        else if (c == '"') {
            std::string str = readString();
            tokens.push_back(Token("string", str, currentLine, currentCol));
        }
        // Handle identifiers and keywords
        else if (isalpha(c) || c == '_') {
            std::string id = readIdentifier();
            std::string type = keywords.count(id) ? "keyword" : "identifier";
            tokens.push_back(Token(type, id, currentLine, currentCol));
        }
        // Handle power operator **
        else if (c == '*' && peek() == '*') {
            advance(); advance();
            tokens.push_back(Token("operator", "**", currentLine, currentCol));
        }
        // Handle shift operators << >>
        else if ((c == '<' || c == '>') && peek() == c) {
            std::string op(2, c);
            advance(); advance();
            tokens.push_back(Token("operator", op, currentLine, currentCol));
        }
        // Handle two-character operators (==, !=, <=, >=, &&, ||)
        else if (std::string("=<>!&|").find(c) != std::string::npos && peek() == '=') {
            std::string op = std::string(1, c) + std::string(1, peek());
            advance(); advance();
            tokens.push_back(Token("operator", op, currentLine, currentCol));
        }
        // Handle logical operators && ||
        else if ((c == '&' && peek() == '&') || (c == '|' && peek() == '|')) {
            std::string op(2, c);
            advance(); advance();
            tokens.push_back(Token("operator", op, currentLine, currentCol));
        }
        // Handle single-character operators
        else if (std::string("+-*/%<>=!&|^~").find(c) != std::string::npos) {
            std::string op(1, c);
            tokens.push_back(Token("operator", op, currentLine, currentCol));
            advance();
        }
        // Handle punctuation including colon
        else if (std::string("(){}[],:;.").find(c) != std::string::npos) {
            std::string punc(1, c);
            tokens.push_back(Token("punctuation", punc, currentLine, currentCol));
            advance();
        }
        else {
            advance();
        }
    }
    return tokens;
}

File: Ollangc/src/Core/Lexer.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace Ollang.Lexer
{
    public class Token
    {
        public string Type;
        public string Value;
        public int Line;
        public int Column;

        public Token(string type, string value, int line = 0, int column = 0)
        {
            Type = type;
            Value = value;
            Line = line;
            Column = column;
        }
    }

    public class LexerState
    {
        private readonly string _src;
        private int _pos = 0;
        private int _line = 1;
        private int _column = 1;

        private static readonly HashSet<string> _keywords = new HashSet<string>
        {
            "var", "func", "if", "else", "while", "return", "true", "false", "null",
            "for", "import", "in", "and", "or", "break", "continue", "try", "catch",
            "finally", "throw", "class", "self", "new", "switch", "case", "do",
            "static", "bool", "num", "str", "void", "int", "float", "double",
            "string", "any", "object", "as", "const", "default"
        };

        private static readonly string[] _multiOps =
        {
            "==", "!=", "<=", ">=",
            "&&", "||", "**", "<<", ">>", "^^"
        };

        private static readonly HashSet<char> _singleOpChars = new HashSet<char>
        {
            '+', '-', '*', '/', '%', '=', '<', '>', '!', '&', '|', '^', '~'
        };

        public LexerState(string source) => _src = source;

        private char Current => _pos < _src.Length ? _src[_pos] : '\0';
        private char Peek(int offset) => _pos + offset < _src.Length ? _src[_pos + offset] : '\0';

        private void Advance(int count = 1)
        {
            for (int i = 0; i < count; i++)
            {
                if (Current == '\n')
                {
                    _line++;
                    _column = 1;
                }
                else
                    _column++;
                _pos++;
            }
        }

        public List<Token> Tokenize()
        {
            var tokens = new List<Token>();

            while (_pos < _src.Length)
            {
                if (char.IsWhiteSpace(Current))
                {
                    Advance();
                    continue;
                }

                if (Current == '/' && Peek(1) == '/')
                {
                    while (Current != '\n' && Current != '\0') Advance();
                    continue;
                }

                if (Current == '/' && Peek(1) == '*')
                {
                    Advance(2);
                    while (!(_pos >= _src.Length || (Current == '*' && Peek(1) == '/')))
                        Advance();
                    if (_pos < _src.Length) Advance(2);
                    continue;
                }

                if (Current == '#')
                {
                    while (Current != '\n' && Current != '\0') Advance();
                    continue;
                }

                int startLine = _line;
                int startCol = _column;

                if (char.IsLetter(Current) || Current == '_')
                {
                    var sb = new StringBuilder();

                    while (char.IsLetterOrDigit(Current) || Current == '_')
                    {
                        sb.Append(Current);
                        Advance();
                    }

                    string value = sb.ToString();
                    tokens.Add(new Token(_keywords.Contains(value) ? "keyword" : "id", value, startLine, startCol));
                    continue;
                }

                if (char.IsDigit(Current))
                {
                    var sb = new StringBuilder();
                    bool hasDot = false;

                    if (Current == '0' && (Peek(1) == 'x' || Peek(1) == 'X'))
                    {
                        sb.Append(Current); Advance();
                        sb.Append(Current); Advance();
                        while (IsHexDigit(Current))
                        {
                            sb.Append(Current);
                            Advance();
                        }
                    }
                    else
                    {
                        while (char.IsDigit(Current) || (Current == '.' && !hasDot && char.IsDigit(Peek(1))))
                        {
                            if (Current == '.') hasDot = true;
                            sb.Append(Current);
                            Advance();
                        }
                    }

                    tokens.Add(new Token("num", sb.ToString(), startLine, startCol));
                    continue;
                }

                if (Current == '"')
                {
                    Advance();
                    var sb = new StringBuilder();

                    while (Current != '"' && Current != '\0')
                    {
                        if (Current == '\\')
                        {
                            Advance();
                            switch (Current)
                            {
                                case 'n': sb.Append('\n'); break;
                                case 'r': sb.Append('\r'); break;
                                case 't': sb.Append('\t'); break;
                                case '"': sb.Append('"'); break;
                                case '\\': sb.Append('\\'); break;
                                case '0': sb.Append('\0'); break;
                                default: sb.Append(Current); break;
                            }
                            Advance();
                        }
                        else
                        {
                            sb.Append(Current);
                            Advance();
                        }
                    }

                    if (Current == '"') Advance();
                    else throw new Exception($"Lexer error at {startLine}:{startCol}: Unterminated string literal");
                    tokens.Add(new Token("str", sb.ToString(), startLine, startCol));
                    continue;
                }

                if (Current == '\'')
                {
                    Advance();
                    var sb = new StringBuilder();
                    while (Current != '\'' && Current != '\0')
                    {
                        if (Current == '\\')
                        {
                            Advance();
                            switch (Current)
                            {
                                case 'n': sb.Append('\n'); break;
                                case 'r': sb.Append('\r'); break;
                                case 't': sb.Append('\t'); break;
                                case '\'': sb.Append('\''); break;
                                case '\\': sb.Append('\\'); break;
                                default: sb.Append(Current); break;
                            }
                            Advance();
                        }
                        else
                        {
                            sb.Append(Current);
                            Advance();
                        }
                    }
                    if (Current == '\'') Advance();
                    else throw new Exception($"Lexer error at {startLine}:{startCol}: Unterminated character literal");
                    tokens.Add(new Token("str", sb.ToString(), startLine, startCol));
                    continue;
                }

                bool matched = false;

                foreach (var op in _multiOps)
                {
                    bool match = true;
                    for (int i = 0; i < op.Length; i++)
                    {
                        if (Peek(i) != op[i])
                        {
                            match = false;
                            break;
                        }
                    }

                    if (match)
                    {
                        tokens.Add(new Token("op", op, startLine, startCol));
                        Advance(op.Length);
                        matched = true;
                        break;
                    }
                }

                if (!matched)
                {
                    tokens.Add(new Token(
                        _singleOpChars.Contains(Current) ? "op" : "punc",
                        Current.ToString(),
                        startLine,
                        startCol
                    ));
                    Advance();
                }
            }

            return tokens;
        }

        private static bool IsHexDigit(char c) =>
            (c >= '0' && c <= '9') || (c >= 'a' && c <= 'f') || (c >= 'A' && c <= 'F');
    }
}